server/infoextractor.js

```diff
@@ -0,0 +1,271 @@
+  }
+}
+
+function getCachedSearchResults(service, query) {
+  return new Promise((resolve, reject) => {
+    redisClient.get(`search:${service}:${query}`, (err, value) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+      if (!value) {
+        resolve(null);
+        return;
+      }
+
+      resolve(JSON.parse(value));
+    });
+  });
+}
+
+function cacheSearchResults(service, query, results) {
+  redisClient.set(`search:${service}:${query}`, JSON.stringify(results), "EX", 60 * 60 * 24, (err) => {
+    if (err) {
+      log.error(`Failed to cache search results: ${err}`);
+    }
+  });
+}
+
+/**
+ * Returns the adapter instance for a given service name.
+ * @param {string} service
+ * @returns {ServiceAdapter}
+ */
+function getServiceAdapter(service) {
+  return adapters.find(adapter => adapter.serviceId === service);
+}
+
+/**
+ * Returns the adapter that can handle a given URL.
+ * @param {string} url
+ * @returns {ServiceAdapter}
+ */
+function getServiceAdapterForURL(url) {
+  return adapters.find(adapter => adapter.canHandleURL(url));
+}
+
+/**
+ * Returns metadata for a single video. Uses cached info if possible and writes newly fetched info
+ * to the cache.
+ * @param {string} service
+ * @param {string} videoId
+ * @returns {Video}
+ */
+async function getVideoInfo(service, videoId) {
+  const adapter = getServiceAdapter(service);
+  const [cachedVideo, missingInfo] = await getCachedVideo(service, videoId);
+
+  if (missingInfo.length === 0) {
+    return cachedVideo;
+  }
+  else {
+    log.warn(`MISSING INFO for ${cachedVideo.service}:${cachedVideo.id}: ${missingInfo}`);
+
+    try {
+      const fetchedVideo = await adapter.fetchVideoInfo(cachedVideo.id, missingInfo);
+      const video = Video.merge(cachedVideo, fetchedVideo);
+      if (adapter.isCacheSafe) {
+        updateCache(video);
+      }
+      return video;
+    }
+    catch (e) {
+      if (e instanceof OutOfQuotaException) {
+        log.error("Failed to get video info: Out of quota");
+        if (missingInfo.length < storage.getVideoInfoFields(cachedVideo.service).length) {
+          log.warn(`Returning incomplete cached result for ${cachedVideo.service}:${cachedVideo.id}`);
+          return cachedVideo;
+        }
+        else {
+          throw e;
+        }
+      }
+      else {
+        log.error(`Failed to get video info for ${cachedVideo.service}:${cachedVideo.id}: ${e}`);
+        throw e;
+      }
+    }
+  }
+}
+
+async function getManyVideoInfo(videos) {
+  const grouped = _.groupBy(videos, "service");
+  const results = await Promise.all(Object.entries(grouped).map(async ([service, serviceVideos]) => {
+    // Handle each service separately
+    const cachedVideos = await storage.getManyVideoInfo(serviceVideos);
+    const requests = cachedVideos
+      .map(video => ({
+        id: video.id,
+        missingInfo: storage.getVideoInfoFields(video.service).filter(p => video[p] == null),
```

Comment:
```suggestion
        missingInfo: storage.getVideoInfoFields(video.service).filter(p => video[p] === null),
```

Reply:
The `== null` and `!= null` checks are intended to be double-equals because they cover both `null` and `undefined` cases. It's actually quite common to do this afaik. eslint even has a rule to allow double-equals comparisons with `null`.